File: agent-driven-architecture-main/examples/kotlin/block/artifact/src/main/kotlin/adr/blocks/artifact/Fold.kt
// ── blocks/artifact/fold — the block's ARM ─────────────────────────────────
// One closed match over SealStatus, no else arm. Every line the work product will
// ever contain is written HERE, by a pure function of committed state — which is
// what makes 2.2's "the folded, replayable result of the session" true.

package adr.blocks.artifact

import adr.contract.ArtifactEffect
import adr.contract.ArtifactResult
import adr.contract.ArtifactResult.ArtifactLine
import adr.spine.pure.ArmOut
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp

internal class ArtifactArm {

    fun arm(
        slice: ArtifactSlice,
        result: ArtifactResult,
        now: Timestamp,
        sig: Signature,
    ): ArmOut<ArtifactSlice> = when (result) {
        is ArtifactResult.RecordFinding -> when (slice.seal) {
            // A sealed artifact is closed: appending to it would change what was delivered.
            is SealStatus.Sealed -> reject(slice, now, result, "the work product is already sealed")
            SealStatus.Draft, is SealStatus.Sealing ->
                if (result.text.isBlank()) {
                    reject(slice, now, result, "a finding cannot be blank")
                } else {
                    ArmOut(slice = slice.withLine(ArtifactLine(now, sig.by, result.text)))
                }
        }

        is ArtifactResult.RequestSeal -> when (slice.seal) {
            SealStatus.Draft ->
                // Reversible: a request emits NO effect.
                ArmOut(slice = slice.withSeal(SealStatus.Sealing(sig.authority)))

            is SealStatus.Sealing -> reject(slice, now, result, "a seal has already been requested")
            is SealStatus.Sealed -> reject(slice, now, result, "the work product is already sealed")
        }

        is ArtifactResult.ConfirmSeal -> when (slice.seal) {
            is SealStatus.Sealing ->
                // RULE 2: the one irreversible effect lives INSIDE the success branch.
                ArmOut(
                    slice = slice.withSeal(SealStatus.Sealed(now, sig.authority)),
                    effects = listOf(ArtifactEffect.DeliverArtifact(now, slice.lines)),
                )

            SealStatus.Draft -> reject(slice, now, result, "no seal has been requested")
            is SealStatus.Sealed -> reject(slice, now, result, "the work product is already sealed")
        }
    }

    private fun reject(
        slice: ArtifactSlice,
        now: Timestamp,
        result: ArtifactResult,
        reason: String,
    ): ArmOut<ArtifactSlice> =
        ArmOut(slice = slice, notices = listOf(Notice.Rejected(now, result.tool, reason)))
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/artifact/fold.ts
// ── blocks/artifact/fold — one line per arm (G16) ──────────────────────────
// 13.1 step 6 used to "perform the delivery effect that writes one line to the
// work product". It folds one line now. Delivery moved to seal time and became
// a single gated irreversible effect.
//
// `ArtifactLine.by` is `sig.by` — the boundary's stamp, arriving with the fold.
// It is NOT something a tool put in its payload, which is exactly the
// distinction G1 exists to enforce.

import type { Signature } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";
import { rejected } from "@adr/spine/pure/notice";
import type { ArmOut } from "@adr/spine/pure/verb";
import { armOut } from "@adr/spine/pure/verb";
import type { ArtifactResult, DeliverArtifact } from "./contract";
import type { ArtifactSlice, SealStatus } from "./slice";
import { sealed, sealing, withLine, withSeal } from "./slice";

export function artifactArm(
  slice: ArtifactSlice,
  r: ArtifactResult,
  now: Timestamp,
  sig: Signature,
): ArmOut<ArtifactSlice> {
  switch (r.tool) {
    case "recordFinding": {
      if (!isOpen(slice.seal)) {
        return armOut(slice, [], [rejected(now, r.tool, "the work product is already sealed")]);
      }
      return armOut(withLine(slice, { at: now, by: sig.by, text: r.text }), [], []);
    }
    case "requestSeal": {
      if (!isOpen(slice.seal)) {
        return armOut(slice, [], [rejected(now, r.tool, "the work product is already sealed")]);
      }
      // a request is REVERSIBLE — nothing is delivered here
      return armOut(withSeal(slice, sealing(sig.authority)), [], []);
    }
    case "confirmSeal": {
      // reads the PARENT-declared property, so the seal status can grow a
      // variant without adding a match to this arm
      if (slice.seal.requestedBy === null) {
        return armOut(slice, [], [rejected(now, r.tool, "no pending seal request")]);
      }
      // THE ONE PINNED CONSTRUCTION SITE for this leaf (check C17).
      const delivery: DeliverArtifact = {
        kind: "DeliverArtifact",
        at: now,
        effectClass: "Irreversible",
        lines: slice.lines,
      };
      return armOut(withSeal(slice, sealed(now, sig.authority)), [delivery], []);
    }
    default: {
      const _never: never = r;
      return _never;
    }
  }
}

// The block's ONE closed match over SealStatus.
function isOpen(seal: SealStatus): boolean {
  switch (seal.kind) {
    case "Draft":
      return true;
    case "Sealing":
      return true;
    case "Sealed":
      return false;
    default: {
      const _never: never = seal;
      return _never;
    }
  }
}
